pkg/rbd/nodeserver.go

```diff
@@ -248,21 +299,75 @@ func (ns *NodeServer) NodeUnpublishVolume(ctx context.Context, req *csi.NodeUnpu
 		return nil, status.Error(codes.NotFound, err.Error())
 	}
 	if notMnt {
-		// TODO should consider deleting path instead of returning error,
-		// once all codes become ready for csi 1.0.
-		return nil, status.Error(codes.NotFound, "volume not mounted")
+		if err = os.RemoveAll(targetPath); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &csi.NodeUnpublishVolumeResponse{}, nil
+	}
+
+	if err = ns.mounter.Unmount(targetPath); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	if err = os.RemoveAll(targetPath); err != nil {
+		return nil, status.Error(codes.Internal, err.Error())
+	}
+
+	klog.Infof("rbd: successfully unbinded volume %s from %s", req.GetVolumeId(), targetPath)
+
+	return &csi.NodeUnpublishVolumeResponse{}, nil
+}
+
+// NodeUnstageVolume unstages the volume from the staging path
+func (ns *NodeServer) NodeUnstageVolume(ctx context.Context, req *csi.NodeUnstageVolumeRequest) (*csi.NodeUnstageVolumeResponse, error) {
+	var err error
+	if err = util.ValidateNodeUnstageVolumeRequest(req); err != nil {
+		return nil, err
 	}
 
-	devicePath, cnt, err := mount.GetDeviceNameFromMount(ns.mounter, targetPath)
+	stagingTargetPath := req.GetStagingTargetPath()
+
+	// kind of hack to unmount block volumes
+	blockStagingPath := stagingTargetPath + "/" + req.GetVolumeId()
+unmount:
+	notMnt, err := mount.IsNotMountPoint(ns.mounter, stagingTargetPath)
+	if err != nil {
+		if os.IsNotExist(err) {
+			// staging targetPath has already been deleted
+			klog.V(4).Infof("stagingTargetPath: %s has already been deleted", stagingTargetPath)
+			return &csi.NodeUnstageVolumeResponse{}, nil
+		}
+		return nil, status.Error(codes.NotFound, err.Error())
+	}
+
+	if notMnt {
+		_, err = os.Stat(blockStagingPath)
+		if err == nil && (stagingTargetPath != blockStagingPath) {
+			stagingTargetPath = blockStagingPath
+			goto unmount
+		}
+		if err = os.RemoveAll(stagingTargetPath); err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+		return &csi.NodeUnstageVolumeResponse{}, nil
+	}
+	// Unmount the volume
+	devicePath, cnt, err := mount.GetDeviceNameFromMount(ns.mounter, stagingTargetPath)
```

Comment:
Code from here on seems to do a few things that may need a revisit with the above hack. For example should we bother about the `cnt` as CO should stage once and unstage it, should it bother us that there are more mounts using the device? Anyway, possibly an issue filed to take a relook at this with the above is required. What are your thoughts?

Reply:
staging should be done only once per PVC on a perticular node as the staging path is derived from PVC name.


> Code from here on seems to do a few things that may need a revisit with the above hack. For example should we bother about the `cnt` as CO should stage once and unstage it, should it bother us that there are more mounts using the device? Anyway, possibly an issue filed to take a relook at this with the above is required. What are your thoughts?

can we end up in such a situation? if yes let's file an issue and look at it

Reply:
I suggest we open an issue to keep track of the hack and this issue once we merge this code, so that we do not lose sight of the same.